### PR TITLE
updated the App.js to avoid merge conflict and created a CartContextW…

### PR DESCRIPTION
**Pull request by joybiswas704**
…rapper component to separate the context from App component

### CODE REVIEW

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+import { CartContext } from "../../util/context";
+
+export const CartContextWrapper = ({ children }) => {
+  const [cart, setCart] = useState([
+    {
+      id: 1,
+      unitPrice: 1200,
+      quantity: 1,
+      maxQuantity: 5,
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:11]: I thought we agreed on having only `id, price, and quantity` in the cart?

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      imageURL: "https://bit.ly/2Z4KKcF",
+      title: "Product Title 4",
+      description: "Lorem ipsum dolor sit amet",
+    },
+  ]);
+
+  const calculateTotalItemsInCart = () => {
+    let totalItem = 0;
+    try {
+      cart.forEach((product) => {
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:51]: Let's use `.map()` for consistency with the other code.

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        totalItem = totalItem + product.quantity;
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:52]: What if the `quantity` happens to be a string?

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 2 above)
Comment (saifulislamdev) [reply src/components/CartContextWrapper/CartContextWrapper.js:51]: + 2 other uses of this method in this file.

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 3 above)
Comment (saifulislamdev) [reply src/components/CartContextWrapper/CartContextWrapper.js:52]: + 2 other situations where we need to take care of the case where quantity is a string.

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      console.log(error);
+    }
+  };
+
+  const calculateTotalPriceInCart = () => {
+    let totalPrice = 0;
+    try {
+      cart.forEach((product) => {
+        totalPrice =
+          totalPrice + product.quantity * parseFloat(product.unitPrice);
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:65]: What if the `unitPrice` is a string?

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+      });
+      return totalPrice.toFixed(2);
+    } catch (error) {
+      console.log(error);
+    }
+  };
+
+  const handleAddQuantity = (id, newQuantity) => {
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:73]: Where is the method for adding a new item to the cart?

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 2 above)
Comment (joybiswas704) [reply src/components/CartContextWrapper/CartContextWrapper.js:51]: map() function returns an array. i need to change element inside the existing array thats why i didnt use map

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 1 above)
Comment (joybiswas704) [reply src/components/CartContextWrapper/CartContextWrapper.js:11]: i added maxQuantity to add a feature on the button where customer can not add more than the quantity. But i will remove it

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 3 above)
Comment (joybiswas704) [reply src/components/CartContextWrapper/CartContextWrapper.js:52]: Thats a good point.  I will add parseInt to type cast it

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 4 above)
Comment (joybiswas704) [reply src/components/CartContextWrapper/CartContextWrapper.js:65]: Thats why i manually type casted to float

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 5 above)
Comment (joybiswas704) [reply src/components/CartContextWrapper/CartContextWrapper.js:73]: i didnt implement it since the product browse page wasnt implemented.

Comment (joybiswas704) [reply src/components/CartContextWrapper/CartContextWrapper.js:73]: I will add it anyway

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 2 above)
Comment (zealptl) [reply src/components/CartContextWrapper/CartContextWrapper.js:47]: you can use reduce method to sum all prices like 
`products.reduce((prev, curr) => prev + curr.price, 0)`

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 4 above)
Comment (zealptl) [reply src/components/CartContextWrapper/CartContextWrapper.js:61]: price field should always be of type number so it's not necessary to type cast, but doesn't hurt to do

File: src/components/App.js [hunk 6]
```diff
@@ -1,51 +1,58 @@
-import { ChakraProvider } from '@chakra-ui/react';
-import { Container } from '@chakra-ui/react';
-import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
+import { ChakraProvider } from "@chakra-ui/react";
```
Comment (saifulislamdev) [inline src/components/App.js:1]: Still using double quotes.

File: src/components/Cart/Cart.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
+import React, { useContext } from "react";
```
Comment (saifulislamdev) [inline src/components/Cart/Cart.js:1]: Hard to follow imports. Not necessary for now, but try to maintain imports in a certain order. The following order is what I am following:
1. React-related imports
2. Chakra-related imports
3. Custom component imports
4. Custom configuration file imports

File: src/components/Cart/Cart.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                  />
+                ))}
+              </Box>
+            </GridItem>
+            <GridItem colSpan={1}>
+              <VStack>
+                <OrderSummary checkout={true} />
+                <Text>
+                  Or,{" "}
+                  <Link
```
Comment (saifulislamdev) [inline src/components/Cart/Cart.js:64]: You could have also named this to `ChakraLink` for added context.

File: src/components/Cart/Cart.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+                    as={RouterLink}
+                    color={colors.primary}
+                    to='/product/product-listing'
```
Comment (saifulislamdev) [inline src/components/Cart/Cart.js:67]: We discussed that this should be changed.

File: src/components/Cart/Cart.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+                  </Link>
+                </Text>
+              </VStack>
+            </GridItem>
+          </SimpleGrid>
+        </Flex>
+      </Elements>
+    </div>
+  );
+};
```
Comment (saifulislamdev) [inline src/components/Cart/Cart.js:79]: Overall, some minor things in this file. For the files where you see double quotes, try to change to single quotes by updating your configuration for Prettier. Also, please keep in mind organization so that it's easy for the eye to read code. It will be easier to read if there is a repeated organization (this includes how you pass in props to a comment).

File: src/components/Cart/CartItem.js [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+      maxW='full'
+      borderWidth='1px'
+      borderRadius='16px'
+      p={4}
+      mb={6}
+      align='center'
+    >
+      <Image
+        src={imageURL}
+        alt='image'
```
Comment (saifulislamdev) [inline src/components/Cart/CartItem.js:52]: In regards to best practices, it is best to include a descriptive alt tag. It can be as simple as "Image of 2022 MacBook Air" (including the item's title is good enough).

File: src/components/Cart/CartItem.js [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+      />
+      <Box p={4}>
+        <Heading size='l'>{title}</Heading>
+        <Text>{description}</Text>
+      </Box>
+      <Box p={4}>
+        <NumberInput
+          defaultValue={0}
+          min={1}
+          onChange={(value) => HandleQuantityChange(value)}
```
Comment (saifulislamdev) [inline src/components/Cart/CartItem.js:65]: Very minor: improper naming of `HandleQuantityChange` function (should be `handleQuantityChange`). Other instances of an improved organization are in this file as well.

File: src/components/Cart/CartItem.js [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+          size='sm'
+        >
+          <NumberInputField maxW={32} />
+          <NumberInputStepper>
+            <NumberIncrementStepper />
+            <NumberDecrementStepper />
+          </NumberInputStepper>
+        </NumberInput>
+      </Box>
+      <Spacer />
```
Comment (saifulislamdev) [inline src/components/Cart/CartItem.js:76]: Nice! Eliminate the use of additional styling by utilizing what Chakra gives us.

File: src/components/Checkout/OrderSummary.js [hunk 14]
```diff
@@ -1,163 +1,179 @@
...
-		</VStack>
-	);
+  return (
+    <VStack
+      w='full'
+      p={10}
+      spacing={10}
+      alignItems='flex-start'
+      style={{
+        border: "2px solid #E2E8F0",
```
Comment (saifulislamdev) [inline src/components/Checkout/OrderSummary.js:100]: You can use Chakra's border and borderRadius props to do this instead of overriding styles.

File: src/components/Checkout/OrderSummary.js [hunk 15]
```diff
@@ -1,163 +1,179 @@
...
+          <Text>{`$ 00.00`}</Text>
+        </GridItem>
+        <GridItem colSpan={2}>
+          <Divider />
+        </GridItem>
+        <GridItem colSpan={1}>
+          <Heading size='l'>Order total:</Heading>
+        </GridItem>
+        <GridItem colSpan={1}>
+          <Heading size='l'>{`$ 00.00`}</Heading>
```
Comment (saifulislamdev) [inline src/components/Checkout/OrderSummary.js:131]: I know these are placeholders, but there is no need for space between $ and 0

File: src/components/Checkout/OrderSummary.js [hunk 16]
```diff
@@ -1,163 +1,179 @@
...
+        </GridItem>
+        <GridItem colSpan={2}>
+          {checkout ? (
+            <Link to='/checkout'>
+              <Button size='md' w='full' bg={colors.primary} color='white'>
```
Comment (saifulislamdev) [inline src/components/Checkout/OrderSummary.js:136]: Using `colorScheme` might add a hover effect.

File: src/components/Checkout/OrderSummary.js [hunk 17]
```diff
@@ -1,163 +1,179 @@
...
+                Checkout
+              </Button>
+            </Link>
+          ) : (
+            <Button
+              disabled={isLoading || !stripe || !elements}
```
Comment (saifulislamdev) [inline src/components/Checkout/OrderSummary.js:142]: Add `colorScheme` as well.

File: src/components/Checkout/OrderSummary.js [hunk 18]
```diff
@@ -1,163 +1,179 @@
...
+              size='md'
+              w='full'
+              bg={colors.primary}
+              color='white'
+              onClick={handleSubmit}
+            >
+              <span id='button-text'>
```
Comment (saifulislamdev) [inline src/components/Checkout/OrderSummary.js:149]: This isn't the only button on the page; it's better to use a more descriptive name.

File: src/components/Checkout/OrderSummary.js [hunk 19]
```diff
@@ -1,163 +1,179 @@
...
+                  "Place Order"
+                )}
+              </span>
+            </Button>
+          )}
+        </GridItem>
+        <GridItem colSpan={2}>
+          {message && (
+            <div
+              style={{
```
Comment (saifulislamdev) [inline src/components/Checkout/OrderSummary.js:162]: No need to override styles as these styles are provided by Chakra.

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 20]
```diff
@@ -0,0 +1,124 @@
+import React, { useState } from "react";
+
+import { CartContext } from "../../util/context";
+
+export const CartContextWrapper = ({ children }) => {
+  const [cart, setCart] = useState([
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:6]: Let's eliminate the initial cart items so that we don't have to remove these later on.

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 21]
```diff
@@ -0,0 +1,124 @@
...
+
+  const calculateTotalItemsInCart = () => {
+    let totalItem = 0;
+    try {
+      cart.forEach((product) => {
+        totalItem = totalItem + parseInt(product.quantity);
+      });
+      return totalItem;
+    } catch (error) {
+      console.log(error);
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:52]: Is there something else you can think of in the catch block beside a log to console? We can leave it like this if there's nothing else.

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+      return totalPrice.toFixed(2);
+    } catch (error) {
+      console.log(error);
+    }
+  };
+
+  const handleAddQuantity = (id, newQuantity) => {
+    try {
+      let newCart = [...cart];
+      newCart.forEach((product) => {
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:72]: Instead, you can do something similar to:
```javascript
setCart((cart) => cart.ForEach((product) => { // the rest of your code
```

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+        if (product.id === id) {
+          product.quantity = newQuantity;
+        }
+      });
+      setCart(newCart);
+    } catch (error) {
+      console.log(error);
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:79]: Again, can we think of something else to do here? If not, that is fine.

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 24]
```diff
@@ -0,0 +1,124 @@
...
+    }
+  };
+
+  const handleAddToCart = ({
+    id,
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:84]: Since id can get you the unitPrice, imageURL, title, and description, we don't need to pass in these arguments. For this function, we can simply pass in id and quantity.

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 25]
```diff
@@ -0,0 +1,124 @@
...
+  }) => {
+    const newProduct = {
+      id,
+      unitPrice,
+      quantity,
+      imageURL,
+      title,
+      description,
+    };
+    const newCart = [...cart, newProduct];
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:99]: Or `setCart((cart) => cart.push(newProduct)`?

File: src/components/CartContextWrapper/CartContextWrapper.js [hunk 26]
```diff
@@ -0,0 +1,124 @@
...
+    console.log(newCart);
+    setCart(newCart);
+  };
+
+  const handleRemoveItemFromCart = (id) => {
+    const newCart = cart.filter((product) => product.id !== id);
```
Comment (saifulislamdev) [inline src/components/CartContextWrapper/CartContextWrapper.js:105]: Nice!

**Review by saifulislamdev — CHANGES_REQUESTED** [review]
I left some comments. Try to emphasize the organization as well.

File: src/components/App.js (hunk 6 above)
Comment (joybiswas704) [reply src/components/App.js:1]: done

File: src/components/Cart/Cart.js (hunk 7 above)
Comment (joybiswas704) [reply src/components/Cart/Cart.js]: done

File: src/components/Cart/Cart.js (hunk 9 above)
Comment (zealptl) [reply src/components/Cart/Cart.js:68]: ^ like saiful said just instead of setting color you can set colorScheme to 'teal'. Pretty sure there's a colors.colorScheme property which you can use

File: src/components/CartContextWrapper/CartContextWrapper.js (hunk 23 above)
Comment (zealptl) [reply src/components/CartContextWrapper/CartContextWrapper.js:53]: take a look at this stackover sol

```
setDatas(
    datas.map(item => 
        item.id === index 
        ? {...item, someProp : "changed"} 
        : item 
))
```

https://stackoverflow.com/questions/55987953/how-do-i-update-states-onchange-in-an-array-of-object-in-react-hooks

File: src/components/Cart/Cart.js (hunk 9 above)
Comment (saifulislamdev) [reply src/components/Cart/Cart.js:68]: For the colorScheme, you can use `colors.colorScheme`, but also the value of `to` needs to be changed.
